semgrep-core/src/engine/Eval_generic.ml

```diff
@@ -158,9 +158,20 @@ let rec eval env code =
       match v with
       | Int _ -> v
       | String s -> (
-          match int_of_string_opt s with
-          | None -> raise (NotHandled code)
-          | Some i -> Int i)
+          (* This code is meant to emulate int(string) in Python.
+           * But! There is one difference:
+           * when called as int("12px"), it will still return 12 instead of erroring.
```

Comment:
What the spec of `metavariable-comparison` says?

> The metavariable-comparison operator compares metavariables against a basic [Python comparison](https://docs.python.org/3/reference/expressions.html#comparisons) expression.

This breaks that spec and it will produce extra findings... and some users may not be happy about that. Rather than changing what `int()` does, I would prefer adding a different function `r2c_int` or something.

Ideally we should just expand the subset of Python that we support. But if that is a hassle, then I think the second best solution is to add `r2c_xyz` functions.

Reply:
Or, maybe, we pretend there is an `r2c` module being imported, and you can refer to the functions as `r2c.my_int`?

Reply:
Good point! Thanks for bringing this up, I wasn't sure what level of faithfulness we were looking for but this makes perfect sense. Since it's common in python to define helper functions without a namespace, I didn't bother with a virtual `r2c.` module. Also, I figured it's more clean two have two composable functions instead of a regular `int(...)` and a reimplementation of `int(...)` with different behavior. So I ended up adding a `digits(value: str) -> str` function and documented it like this: https://github.com/returntocorp/semgrep-docs/pull/453/files#diff-27ce63caf629155835158f784d2556151d037d7018d5618da4eab9cce3195162R208-R214

Reply:
What if we added support for Python's `rstrip` ? You can then write your example as `$PIXELS.rstrip("px") > 1`. It's more explicit about what you are expecting (the `px` suffix) and it's pure Python.

Reply:
:one: starting off with a nitpick: the correct methods would be `.removeprefix()` and `.removesuffix()`, since the spec of `rstrip()` says it takes a sequence of characters and it removes all occurrences of any of them from the right of the string. So it'd also convert `"12pxxxpxpxpxppxp"` to `"12"`. And `12xp` to `12`. Neither of those are likely desired (imagine trying to match a distance in meters, but getting `"1000mm".rstrip("m") == "1000"`).

:two: With that aside, our rule syntax spec already has a `strip: true` convenience feature, which I was hoping to just make smarter, instead of making everyone increase the complexity of their Python comparisons. Basically I'd prefer if people could write

```yaml
strip: true
comparison: $X == $Y
```

instead of

```yaml
comparison: int($X.removesuffix("px")) == int($Y.removesuffix("px"))
```

in their rules. We can make `strip` do the right thing with int & digits cause they take no parameters other than the metavar. But if we have `.removesuffix()` instead, that is not possible.

:three: For the font-size case, you can specify the size as both integer values and strings, see excerpt from semgrep.dev code:

<img width="383" alt="CleanShot 2022-02-28 at 08 57 51@2x" src="https://user-images.githubusercontent.com/1060436/156025170-f069b7d5-5b43-4da6-a823-5a52ea9421c8.png">
<img width="339" alt="CleanShot 2022-02-28 at 08 58 09@2x" src="https://user-images.githubusercontent.com/1060436/156025210-5959403b-bed4-49eb-9b2a-147520e5f8d8.png">

If we tried to keep 100% purity, `$PIXELS_INTEGER.removesuffix("px")` would throw an error, since this method does not makes sense and is not available on integers. So you'd end up having to add a `pattern-either`, then typed metavariable matching, and duplicate your metavar-comparison. What's worse is, without an IDE that can show you hints about types, this issue would be near-impossible to find out about and then debug. So considering the lack of tooling available to write correct metavar-comparison snippets… I think we're better off keeping it simple and just handling this for users who turn on `strip: true`.

Reply:
Note that
```
strip: true
comparison: $X == $Y
```
does not work as you suggest, you need to pick _one_ metavariable to _strip_.

Reply:
Another good reason to redesign metavariable-comparison and make strip() (and other useful functions) regular functions you can use on any metavariables in the comparison expression.

Reply:
I've created this new issue: https://github.com/returntocorp/semgrep/issues/4764